packages/react-server/src/ReactFlightServer.js

```diff
@@ -669,6 +675,16 @@ export function resolveModelToJSON(
   key: string,
   value: ReactClientValue,
 ): ReactJSONValue {
+  if (typeof value === 'string' && value[value.length - 1] === 'Z') {
```

Comment:
Let's see if we can find a way not to duplicate this check. We have a string check further below, that can maybe unify with this. Since that is also going to check the first character both operations are free.

However, this technique doesn't seem very proof for other types of values. We don't have immediate plans to support URL but it's often requested - and it has a toJSON too.

Another thing we could do is always operate on the original value. Might make it more future proof. E.g. FormData doesn't have toJSON today but it might in the future. This might be bad for perf since it's a mega-morphic key look up for every value.

Reply:
How's this? ¯\\\_(ツ)\_/¯  LMK if you prefer always reading the original value. (At that point it might be faster to reimplement JSON.stringify?)